packages/block-editor/src/store/reducer.js

```diff
@@ -192,6 +195,154 @@ export function isUpdatingSameBlockAttribute( action, lastAction ) {
 	);
 }
 
+/**
+ * Higher-order reducer intended to reset the cache key of all blocks
+ * whenever the post meta values change.
+ *
+ * @param {Function} reducer Original reducer function.
+ *
+ * @return {Function} Enhanced reducer function.
+ */
+const withPostMetaUpdateCacheReset = ( reducer ) => ( state, action ) => {
+	const newState = reducer( state, action );
+	const previousMetaValues = get( state, [ 'settings', '__experimentalMetaSource', 'value' ] );
+	const nextMetaValues = get( newState.settings.__experimentalMetaSource, [ 'value' ] );
+	// If post meta values change, reset the cache key for all blocks
+	if ( previousMetaValues !== nextMetaValues ) {
+		newState.blocks = {
+			...newState.blocks,
+			cache: mapValues( newState.blocks.cache, () => ( {} ) ),
+		};
+	}
+
+	return newState;
+};
+
+/**
+ * Higher-order reducer intended to compute a cache key for each block in the post.
+ * A new instance of the cache key (empty object) is created each time the block object
+ * needs to be refreshed (for any change in the block or its children).
+ *
+ * @param {Function} reducer Original reducer function.
+ *
+ * @return {Function} Enhanced reducer function.
+ */
+const withBlockCache = ( reducer ) => ( state = {}, action ) => {
+	const newState = reducer( state, action );
+
+	if ( newState === state ) {
+		return state;
+	}
+	newState.cache = state.cache ? state.cache : {};
+
+	const addParentBlocks = ( clientIds ) => {
+		return clientIds.reduce( ( result, clientId ) => {
+			let current = clientId;
+			do {
+				result.push( current );
+				current = state.parents[ current ];
+			} while ( current );
+			return result;
+		}, [] );
+	};
+
+	const fillKeysWithEmptyObject = ( clientIds ) => {
```

Comment:
The functions addParentBlocks, and fillKeysWithEmptyObject don't seem to depend on the scope of withBlockCache, what if we defined this functions outside of it?
The advantages being:
	- When looking at the code we know they don't depend on scope more easily, they just depend on its parameters.
	- Maybe in some JS engines, there is some benefit in defining them outside, as they will not have to created every time withBlockCache is called. I guess there are optimizations for that but still, maybe there is some gain.

Reply:
addParents is dependent on the state but it's true for fillKeysWithEmptyObjects.